src/java/com/zimbra/cs/nginx/NginxLookupExtension.java

```diff
@@ -880,13 +880,23 @@ private DomainExternalRouteInfo getDomainExternalRouteInfo(ILdapContext zlc, Con
 
         private Server lookupUpstreamImapServer(NginxLookupRequest req) throws ServiceException {
             ImapLoadBalancingMechanism LBMech = ImapLoadBalancingMechanism.newInstance();
-            Server server = prov.getLocalServer();
-            String[] imapServerAddrs;
-            if (server != null) {
-                imapServerAddrs = server.getReverseProxyUpstreamImapServers();
-            } else {
-                imapServerAddrs = prov.getConfig().getReverseProxyUpstreamImapServers();
+            String[] imapServerAddrs = null;
+            Account acct = prov.get(AccountBy.name, req.user);
+            if (acct != null) {
+                imapServerAddrs = acct.getServer().getReverseProxyUpstreamImapServers();
```

Comment:
The interesting thing I have found is that if the requesting 'account' doesn't exist this or the password is incorrect this code is not called. It appears to be called only in the case of an actual authenticating account. 

That being said I'd be happy to add guard code anyhow.